go/pkg/io/walker/walker.go

```diff
@@ -0,0 +1,327 @@
+	// This behavior is equivalent to copying the entire tree of the target in place of the symlink.
+	Replace
+
+	// Cancel indicates that the walker should cancel the entire walk.
+	Cancel
+)
+
+// ErrBadNextStep indicates a fatal error due to an invalid next step returned by the client.
+var ErrBadNextStep = errors.New("invalid next step")
+
+// WalkFunc defines the callback signature that clients must specify for walker implementations.
+//
+// Arguments:
+//
+//	realPath is the real path that can be used to access the file.
+//	desiredPath is always identical to realPath unless the traversal is replacing a symlink, in which case it is the desired target path.
+//	info is the result of stating realPath. It is nil for pre-access calls and when err is set.
+//	err is IO error or nil.
+type WalkFunc func(realPath impath.Absolute, desiredPath impath.Absolute, info fs.FileInfo, err error) NextStep
+
+// DepthFirst walks the filesystem tree rooted at the specified root in depth-first-search style traversal.
+// That is, every directory is visited after all its descendants have been visited.
+//
+// The callback is called twice for each path, including root. The first call is made
+// before making any IO calls, and only the path is provided to the function.
+//
+// At this point, the client may instruct the walker to skip the path by returning Skip or to access
+// the path by returning Continue.
+// Alternatively, the client may return Defer to instruct the walker to buffer the path to be visited after all other siblings
+// have been visited.
+// Deferred paths block traversing back up the tree since a parent requires all its children to be processed before itself to honour the DFS contract.
+//
+// The second call is made only if the first call returned Continue and after making an IO call to stat the file.
+// If the stat syscall returns an error, it is passed to the callback to let the client decide whether to skip the path
+// or cancel the entire walk. No other step is allowed, otherwise an ErrBadNextStep is returned and the entire walk is cancelled.
+//
+// If the second call returns Skip and the path is a directory, the walker will skip the directory's tree.
+//
+// Defer is not a valid next step in any second call because the path has already been processed.
+// Replace is only valid for symlinks.
+// An unexpected NextStep value cancels the entire walk.
+//
+// If a symlink is encountered, the callback must return Skip to avoid triggering a recursive walk on the symlink target.
+// Alternatively, returning Replace will instruct the walker to traverse the target as if its path is that of the symlink itself.
+//
+// At any point, the client may return Cancel to cancel the entire walk.
+//
+// The error returned by the walk is either nil or ErrBadNextStep.
+func DepthFirst(root impath.Absolute, exclude Filter, fn WalkFunc) error {
```

Comment:
Does the future use of the walker justify requiring to provide WalkFunc as a parameter instead of defining its behavior within the walker package?

I didn't see the code that uses the walker yet, but looking at the current implementation of visit and processDir I've noticed lots of extra code that handles errors caused by invalid results returned by WalkFunc. Likewise, because of the fact that WalkFunc is called multiple times for the same path and due to the fact that WalkFunc needs to handle various errors, it seems that implementation of the WalkFunc is not trivial and my require deep knowledge of DeepFirst internals to make it right, making the usage of the package cumbersome. 

If it's possible, maybe it'd be simpler from the perspective the caller, but also from the perspective of the walker package to define, if not all, at least some of these behaviors within the package instead of deferring decisions to WalkFunc

Reply:
I have revised the implementation to eliminate runtime errors that can be captured at compile time. I like it better now. Thanks for the comment!